onnxscript/ir/_linked_list.py

```diff
@@ -0,0 +1,291 @@
+        # original_box <=> new_box <=> original_next
+        original_next = box.next
+        box.next = new_box
+        new_box.prev = box
+        new_box.next = original_next
+        original_next.prev = new_box
+
+        # Call the property modifier in case the users want to modify the properties
+        # For example, when a node is added to a graph, we want to set its graph property
+        if property_modifier is not None:
+            property_modifier(new_value)
+
+        # Be sure to update the length
+        self._length += 1
+
+    def remove(
+        self, value: TLinkable, property_modifier: Callable[[TLinkable], None] | None = None
+    ) -> None:
+        """Remove a node from the list."""
+        if value._link_box is None:
+            raise ValueError(f"Value {value!r} does not belong to any list")
+        if value._link_box.owning_list is not self:
+            raise ValueError(f"Value {value!r} is not in the list")
+        # Remove the link box and detach the value from the box
+        value._link_box.erase()
+
+        # Call the property modifier in case the users want to modify the properties
+        # For example, when a node is removed from a graph, we want to unset its graph property
+        if property_modifier is not None:
+            property_modifier(value)
+
+        # Be sure to update the length
+        self._length -= 1
+
+    def append(
+        self, value: TLinkable, property_modifier: Callable[[TLinkable], None] | None = None
+    ) -> None:
+        """Append a node to the list."""
+        self._insert_one_after(self._root.prev, value, property_modifier=property_modifier)
+
+    def extend(
+        self,
+        values: Iterable[TLinkable],
+        property_modifier: Callable[[TLinkable], None] | None = None,
+    ) -> None:
+        for value in values:
+            self.append(value, property_modifier=property_modifier)
+
+    def insert_after(
+        self,
+        value: TLinkable,
+        new_values: Iterable[TLinkable],
+        property_modifier: Callable[[TLinkable], None] | None = None,
+    ) -> None:
+        """Insert new nodes after the given node.
+
+        Args:
+            value: The value after which the new values are to be inserted.
+            new_values: The new values to be inserted.
+            property_modifier: A function that modifies the properties of the new nodes.
+        """
+        if value._link_box is None:
+            raise ValueError(f"Value {value!r} does not belong to any list")
+        if value._link_box.owning_list is not self:
+            raise ValueError(f"Value {value!r} is not in the list")
+        insertion_point = value._link_box
+        for new_value in new_values:
+            self._insert_one_after(
+                insertion_point, new_value, property_modifier=property_modifier
+            )
+            assert new_value._link_box is not None
+            insertion_point = new_value._link_box
+
+    def insert_before(
```

Comment:
nit: possible to call `insert_after` internally?

Reply:
Done